File: assignment1/assignment1-topic-2/FPGrowth.java
import java.util.*;
import java.util.stream.Collectors;


/**
 * FPNode Class
 */
class FPNode {
    String item;
    int count;
    FPNode parent;
    Map<String, FPNode> children;
    FPNode next;

    public FPNode(String item, int count, FPNode parent) {
        this.item = item;
        this.count = count;
        this.parent = parent;
        this.children = new HashMap<>();
        this.next = null;
    }
}

public class FPGrowth {

    /**
     * FP-growth:
     *    1. contruct FP-tree from transactions
     *    2. pattern mining with constructed FP-tree
     * 
     * @param transactions: converted transactions data from csv file
     * @param minSupport: minimum support
     * @return frequent patterns with their frequency
     */
    public static Map<Set<String>, Integer> fpgrowth(Map<Set<String>, Integer> transactions, float minSupport) {
        Object[] result = constructTree(transactions, minSupport);
        // FPNode root = (FPNode) result[0];
        Map<String, Object[]> headerTable = (Map<String, Object[]>) result[1];

        Map<Set<String>, Integer> frequentItemsets = new HashMap<>();
        mineTree(headerTable, minSupport, new HashSet<>(), frequentItemsets);
        return frequentItemsets;
    }

    /**
     * Construct FP-tree with transactions
     * 
     * @param transactions: transactions
     * @param minSupport: minimum support
     * @return FP-tree, Header-table
     */
    private static Object[] constructTree(Map<Set<String>, Integer> transactions, float minSupport) {
        Map<String, Integer> frequencyList = new HashMap<>();

        // add transaction data to headerTable
        transactions.forEach((transaction, freq) -> transaction.forEach(item -> frequencyList.merge(item, freq, Integer::sum)));

        // pruning items which frequency is less than minimum support
        frequencyList.entrySet().removeIf(entry -> entry.getValue() < minSupport);

        // if headerTable is empty, it is impossible to construct FP-tree. Therefore, return {null, null}
        if (frequencyList.isEmpty()) return new Object[]{null, null};

        // generate newHeaderTable which map item -> [freq, FP-node]s
        Map<String, Object[]> newHeaderTable = new HashMap<>();
        frequencyList.forEach((k, v) -> newHeaderTable.put(k, new Object[]{v, null}));

        // generate a root node of FP-tree
        FPNode root = new FPNode(null, 0, null);

        // iterate all transactions and expand FP-tree with each transaction
        transactions.forEach((transaction, freq) -> {

            // each transactions are filtered and sorted with F-list
            List<String> sortedTransaction = transaction.stream()
                    .filter(frequencyList::containsKey)
                    .sorted(Comparator.comparingInt((String item) -> (int) frequencyList.get(item)).reversed())
                    .collect(Collectors.toList());

            // with ordered frequent items updateTree
            FPNode currentNode = root;
            for (String item : sortedTransaction) {
                currentNode = updateTree(item, currentNode, newHeaderTable, freq);
            }
        });
        return new Object[]{root, newHeaderTable};
    }

    /**
     * Update the child node of `node` with item
     * 
     * @param item
     * @param node: current node
     * @param headerTable
     * @param update
     * @return
     */
    private static FPNode updateTree(String item, FPNode node, Map<String, Object[]> headerTable, int update) {

        // if current node contains child item node
        if (node.children.containsKey(item)) {
            node.children.get(item).count += update;
        } else {
            // else, we need to make a new node and linked to current node
            FPNode newNode = new FPNode(item, update, node);
            node.children.put(item, newNode);

            // In addition, connect new node to headerTable
            if (headerTable.get(item)[1] == null) {
                // if item has no nodes, allocate new node as head node
                headerTable.get(item)[1] = newNode;
            } else {
                // else add to tail
                updateHeader((FPNode) headerTable.get(item)[1], newNode);
            }
        }
        return node.children.get(item);
    }

    /**
     * Link targetNode to next to end node of currentNode
     * 
     * @param nodeToTest
     * @param targetNode
     */
    private static void updateHeader(FPNode curretNode, FPNode targetNode) {
        while (curretNode.next != null) {
            curretNode = curretNode.next;
        }
        curretNode.next = targetNode;
    }

    /**
     * Pattern Mining with constructed FP-tree, HeaderTable
     * 
     * @param headerTable
     * @param minSupport
     * @param prefix
     * @param frequentItemsets
     */
    private static void mineTree(Map<String, Object[]> headerTable, float minSupport, Set<String> prefix, Map<Set<String>, Integer> frequentItemsets) {

        // check whether header table is null or not
        if (headerTable == null) return;

        // iterate for all items in headerTable
        headerTable.keySet().forEach(baseItem -> {

            // generate a new prefix pattern
            Set<String> newFrequentSet = new HashSet<>(prefix);
            newFrequentSet.add(baseItem);
            
            // add generated prefix pattern to result set
            frequentItemsets.put(newFrequentSet, (Integer) headerTable.get(baseItem)[0]);

            // make a conditional pattern base with baseItem
            Map<Set<String>, Integer> conditionalPatternBase = findPatternBase(baseItem, headerTable);

            // construct conditional FP-tree with conditional pattern base
            Object[] result = constructTree(conditionalPatternBase, minSupport);

            // FPNode conditionalRoot = (FPNode) result[0];
            Map<String, Object[]> conditionalHeader = (Map<String, Object[]>) result[1];
            mineTree(conditionalHeader, minSupport, newFrequentSet, frequentItemsets);
        });
    }

    /**
     * Find a pattern base of `baseItem`
     * 
     * @param baseItem
     * @param headerTable
     * @return
     */
    private static Map<Set<String>, Integer> findPatternBase(String baseItem, Map<String, Object[]> headerTable) {
        Map<Set<String>, Integer> transactions = new HashMap<>();

        // node is a head node of baseItem
        FPNode node = (FPNode) headerTable.get(baseItem)[1];

        // traverse all nodes which linked to head node of baseItem
        while (node != null) {

            // find a path from node to root
            List<String> path = ascendTree(node);

            // prefixPath[0] is always node, therefore slice it and add to transactions (conditional pattern base)
            if (path.size() > 1) {
                transactions.put(new HashSet<>(path.subList(1, path.size())), node.count);
            }

            // move to next node (same baseItem)
            node = node.next;
        }
        return transactions;
    }

    /**
     * Add a path from `node` to root to `prefixPath`
     * 
     * @param node
     * @param prefixPath 
     */
    private static List<String> ascendTree(FPNode node) {
        List<String> path = new ArrayList<>();
        while (node.parent != null) {
            path.add(node.item);
            node = node.parent;
        }
        return path;
    }

    public static void main(String[] args) {
        Map<Set<String>, Integer> transactions = new HashMap<>();
        transactions.put(new HashSet<>(Arrays.asList("f", "a", "c", "d", "g", "i", "m", "p")), 1);
        transactions.put(new HashSet<>(Arrays.asList("a", "b", "c", "f", "l", "m", "o")), 1);
        transactions.put(new HashSet<>(Arrays.asList("b", "f", "h", "j", "o")), 1);
        transactions.put(new HashSet<>(Arrays.asList("b", "c", "k", "s", "p")), 1);
        transactions.put(new HashSet<>(Arrays.asList("a", "f", "c", "e", "l", "p", "m", "n")), 1);

        float minSupport = 0.05f;
        Map<Set<String>, Integer> patterns = fpgrowth(transactions, minSupport);

        System.out.println("Frequent Itemsets:");
        patterns.forEach((itemset, support) -> System.out.println(itemset + " " + ((float) support / transactions.size())));
    }
}

File: assignment1-topic-2/python-version/FP-growth.py
from collections import defaultdict

# Data Structure for FP-Tree
class FPNode:
    def __init__(self, item, count, parent):
        self.item = item
        self.count = count
        self.parent = parent
        self.children = {}
        self.next = None
        
### FP-growth, it has two steps
#a
# 1. Generate FP-tree
# 2. Mining frequent patterns with FP-tree
        
def fpgrowth(transactions, min_support):
    # 1. Generate FP-tree
    root, header_table = construct_tree(transactions, min_support)
    frequent_itemsets = {}
    
    # 2. Mining frequent patterns with FP-tree
    mine_tree(header_table, min_support, set(), frequent_itemsets)
    return frequent_itemsets
        
        
### GENERATE FP-Tree

# arugments:
#   transactions: List[List[items]]
#   min_support: Int
#
# returns:
#   FP-tree: FPNode
#   header_table: Dict[items]

def construct_tree(transactions, min_support):
    header_table = defaultdict(int)
    
    # get frequency of items
    size = 0
    for transaction, freq in transactions.items():
        for item in transaction:
            header_table[item] += freq
            size += freq
    
    # Remove items below min support
    header_table = {k: v for k, v in header_table.items() if v / size >= min_support}
    frequent_items = set(header_table.keys())
    
    # if no table entries
    if len(frequent_items) == 0:
        return None, None
    
    # preparing for adding FP-nodes to each header
    # header[item][0] = frequency of item
    # header[item][1] = FP-nodes of item
    for k in header_table:
        header_table[k] = [header_table[k], None]
    
    root = FPNode(None, None, None)
    
    # generate FP-tree from transactions
    for transaction, freq in transactions.items():
        
        # get items contained in frequenct_items
        transaction = [item for item in transaction if item in frequent_items]
        
        # sort items with frequency
        transaction.sort(key=lambda item: header_table[item][0], reverse=True)
        
        # following sorted transactions, update tree
        current_node = root
        for item in transaction:
            current_node = update_tree(item, current_node, header_table, update=freq)
    
    return root, header_table

def update_tree(item, node, header_table, update=1):
    if item in node.children:
        # if already node has children
        # just add 1
        node.children[item].count += update
    else:
        # if not
        #   1. make a new FP-node
        #   2. if the node is a first node of item
        #       just connect the node to header[item]
        #   3. if not,
        #       connect the node to tail of header[item][1] -> implemented with `update_header`
        
        # 1.
        node.children[item] = FPNode(item, update, node)
        if header_table[item][1] is None:
            # 2.
            header_table[item][1] = node.children[item]
        else:
            # 3.
            update_header(header_table[item][1], node.children[item])
    return node.children[item]


def update_header(node_to_test, target_node):
    # find tail node
    while node_to_test.next is not None:
        node_to_test = node_to_test.next
        
    # add target node to tail node
    node_to_test.next = target_node



### 2. Mining frequent patterns with FP-tree

def mine_tree(header_table, min_support, prefix, frequent_itemsets):
    for base_item in header_table.keys():
        
        # make a new patterns with prefix + base_item
        new_frequent_set = prefix.copy()
        new_frequent_set.add(base_item)
        frequent_itemsets[frozenset(new_frequent_set)] = header_table[base_item][0]

        # find prefix path with base_item
        conditional_tree = find_prefix_path(base_item, header_table)
        
        # construct conditional FP-tree and FP-growth~
        conditional_root, conditional_header = construct_tree(conditional_tree, min_support)
        if conditional_header is not None:
            mine_tree(conditional_header, min_support, new_frequent_set, frequent_itemsets)
        
            
# calculate conditional FP-graph
def find_prefix_path(base_item, header_table):
    paths = {}
    
    # FP-nodes from base_item
    node = header_table[base_item][1]
    
    while node is not None:
        prefix_path = []
        ascend_tree(node, prefix_path)
        
        # prefix_path[0] == node.item, always
        if len(prefix_path) > 1:
            paths[frozenset(prefix_path[1:])] = node.count
            
        # iterate with next node that has same base_time
        node = node.next
        
    return paths

# generate ascend tree from node to root.
def ascend_tree(node, prefix_path):
    if node.parent is not None:
        prefix_path.append(node.item)
        ascend_tree(node.parent, prefix_path)




# Sample transactions data
transactions = [
    ['milk', 'bread', 'butter'],
    ['milk', 'bread'],
    ['milk', 'eggs'],
    ['bread', 'butter'],
    ['milk', 'bread', 'butter', 'eggs'],
    ['bread', 'eggs']
]

# Sample transcations data in paper
transactions = [
    'facdgimp',
    'abcflmo',
    'bfhjo',
    'bcksp',
    'afcelpmn',
]

# transactions = [['Milk', 'Onion', 'Nutmeg', 'Kidney Beans', 'Eggs', 'Yogurt'],
#            ['Dill', 'Onion', 'Nutmeg', 'Kidney Beans', 'Eggs', 'Yogurt'],
#            ['Milk', 'Apple', 'Kidney Beans', 'Eggs'],
#            ['Milk', 'Unicorn', 'Corn', 'Kidney Beans', 'Yogurt'],
#            ['Corn', 'Onion', 'Onion', 'Kidney Beans', 'Ice cream', 'Eggs']]

transactions_convert = {}
for transaction in transactions:
    transactions_convert[frozenset(transaction)] = transactions_convert.get(frozenset(transaction), 0) + 1


# print(transactions_convert)

# Minimum support threshold
min_support = 0.05

# Perform FP-Growth
patterns = fpgrowth(transactions_convert, min_support)

# Print frequent itemsets
print("Frequent Itemsets:")
for itemset, support in patterns.items():
    print(sorted(itemset), support / len(transactions))
